### DISCUSSION TITLE
0015 XLS-15d: Concise Transaction Identifier

### DISCUSSION THREAD

**Discussion by RichardAH** [XRPLF/XRPL-Standards#34] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by WietseWind** [comment]
Very useful to point eg. with a issued currency HEX code (#30) to a transaction containing memo with more information about the token (@raredata, @nixer89).

A change I'd like to propose is to switch from `xrpl-cti://67835576823535218` to the `xrpl://` protocol code, so apps don't have to register a whole list of protocols to handle, eg.:
`xrpl://cti/67835576823535218`

**Comment by nbougalis** [comment]
This format seems to conflate two different goals: a "compact" encoding for code purposes, and a "compact" encoding for human purposes. Why not use two?

Using your example, here's what it would look like for human consumption:

xrpl://cti/62090589-9
or
xrpl://cti/62090589/9

For the compact 64-bit encoding, useful to machines, I'd recommend 4 bytes for the ledger sequence followed by the 3 low bytes of the sequence number. The last byte can be a CRC8 over the previous 7 bytes, if you're so inclined.

**Reply by RichardAH** [reply]
Main issue with that encoding IMHO is that fails the double-click-copy test. That is: a user in a hurry double clicks (to select) and then copies, resulting in half copied code, whereas a singular code with no delimiter is always copied completely.

**Reply by nbougalis** [reply]
Makes sense, re: copy-paste test. I still think you should use hex: it makes the identifier shorter than pure decimal and more compact for users, which is a plus.

**Reply by RichardAH** [reply]
> Makes sense, re: copy-paste test. I still think you should use hex: it makes the identifier shorter than pure decimal and more compact for users, which is a plus.

I agree hex is a nicer representation for compactness, however decimal is case-less (and therefore every CTI literal is trivially unique, there are no non-canonical CTIs, unlike for example `xrpl://cti/F1001903B35672` and `xrpl://cti/f1001903b35672` which would be two unique strings referring to the same CTI) and can be parsed and passed trivially between virtually any language, system or interface with no special logic. That said the spec leaves it up to the implementer to decide how to display or use a CTI, unless it's a URI, in which case presently decimal is used for the reasons above.

**Comment by Silkjaer** [comment]
Referring to transactions to store e.g. a token payload is great, however it will only live for as long as full history is available. Where “live” ledger objects persistent, the transaction history can be lost, hence invalidating the reference?

It’s not a major concern, as the standards discussed here and in the NFT thread is focusing on ways to implement NFTs without amending the XRPL – so the outcome will be an agreed upon standard of how to do it, and not a “real” XRPL feature.

**Reply by RichardAH** [reply]
It’s definitely true that if the thing you point to disappears the pointer is effectively useless.
But even if that does happen the pointer is still unique. So for example if you did point to a txn containing metadata for an NFT, you lose access to the metadata but you retain the uniqueness of the token. Ultimately the world of crypto currency is a cooperative game where we all agree to assign significance to arbitrary values decided through consensus. Why not the unique pointer being the value as opposed to the pointee?

**Reply by Silkjaer** [reply]
I agree, and a standard for a pointer is very helpful no matter what. Just as a hyperlink remains a hyperlink, even if the domain doesn’t resolve. 

The main concern of mine was related to the NFT-specific use cases and I have addressed that in the proper thread (#30). In other words, the persistence-concern in relation to CTI can be ignored 🙂

**Reply by calvincs** [reply]
Question w/ "Why not the unique pointer being the value as opposed to the pointee?"  If the pointer is failing to resolve to the meta, how does the end users application pull the meta that it requires in order to show the content, or validate they own the digital objects?   This is one thing I really like about the XRP ledger, and I find as one of its strong suites as compared to other block chains, you don't require a full history of TXs.   If we begin to use the meta in the memo of a TX, and then use a pointer which can roll off (unless you maintain full history), you create complexity.  Over time the system will become fragile (NFTs).   

However, if the pointer/meta is in the AccountRoot, no number of transactions is required for storage for later retrieval.
Are my assumptions of this limitation incorrect?

If not, I think we should really consider putting the meta/pointer data in the AccountRoot, like the Domain/Email/Msg Key hash, or create a new field (I dont think this is required tho). 

Im writing a piece of software which may help with this by providing AccountRoot summary data in a p2p fashion for clients.  And while some hosts would need to periodically query the ledger to compile said data, those hosts would not require the full processing power of a full history rippled node to do that, as we are only examining the latest state of the ledgers accounts.  And clients as simple as phone web browsers should be able to acquire, parse, and locate whatever services or meta they require from the ledger, from a given address, in a reasonable time period.

**Reply by calvincs** [reply]
If you can use the strengths the XRP ledger gives you out of the box, and boot strap your apps that way instead of requiring heavy services like full history nodes, IMHO it makes the app and the use cases stronger.   Small light weight apps should have as easy of a time locating services or pointers tied to assets on the ledger as possible w/o requiring resource intensive infrastructure.

I understand Xumm has a well  maintained back end, full history nodes, beefy post processing w/ lookup tables, etc.  But not all apps have that power, and if possible, if we can make that not a dependency for applications, XRP dependent apps and services would be better for it.  This would also allow more agility for apps, less dependence on said heavy services like full history nodes do function normally.  

This is why when I built the demo code I did, it used the accountRoot, the one thing that always carries forward.
Thoughts?

**Reply by RichardAH** [reply]
> However, if the pointer/meta is in the AccountRoot, no number of transactions is required for storage for later retrieval.
> Are my assumptions of this limitation incorrect?


XLS15d is just a definition for a type of pointer called CTI, it doesn't tell you how you should or shouldn't use it. 
I think you probably want to move this NFT discussion to XLS16d which proposes using CTI as a pointer to NFT metadata.

**Comment by intelliot** [comment]
Since the first hex nibble is only affected by the first four bits of ledger hash and the first four bits of transaction hash, it seems there is a risk of a transcription error since it is not a checksum over the entire data. It may make sense to use something like the [Luhn algorithm](https://en.wikipedia.org/wiki/Luhn_algorithm) (which is used in credit card numbers) so that any single digit error is caught.

**Reply by RichardAH** [reply]
The first byte isn’t a checksum over the data so this doesn’t work the way you’re thinking. The address of the transaction is validated with some information about the destination that you can only validate by looking it up / going there.

If I give you street directions and tell you that the house you are looking for has a green dragon out front now I can make any number of transcription errors and you will still know you arrived at the wrong place in almost every iteration of this scenario.

In this case even if the ledger and transaction you ended up accidentally addressing are valid you have only a 1 in 256 chance of getting the check byte correct with respect to the destination. Thus for at least 255 out of 256 mistranscriptions this will catch any number of errors in the digits.

**Reply by intelliot** [reply]
Ah, my bad. I understand now -- thanks!

**Comment by intelliot** [comment]
Alternatively, [Bech32](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki) is a little more compact.

**Comment by ximinez** [comment]
> We assume a ledger will never contain more than 65535 transactions.

I don't know if that's a very safe assumption. Sure, we're nowhere near that level _today_, but there's no telling how things may change in the future.

**Reply by RichardAH** [reply]
That’s true, however this would gracefully fail in the form of CTI-addressable vs not CTI addressable txns. The cost of using two extra bytes everywhere forever for the possibility of not being able to address some txns in the future seems unappealing to me. I suppose “zero compression” could be used but adds unwanted complexity to a currently straight forward standard

**Reply by ximinez** [reply]
Good points. This standard 1) is very clear that it has this restriction, and 2) leaves open the possibility of coming up with a new standard if and when ledgers ever get that big. Sort of a T65K problem, if you will. :grinning:

**Comment by nbougalis** [comment]
Having now read this spec and the several comments, I agree that the 64k limitation on transactions per ledger is likely a non-issue today but "feels" problematic nonetheless. Since we're still discussing this, let's weigh the positives of a _slightly_ more complex format: 

### Proposed alternative

Part of my goal was to ensure that any tx which has a valid CTI under the original proposal (i.e. has a 64-bit encoding) the resulting encoding under this scheme would _also_ be under 64 bits. In other words, we want to fit the resulting CTI in 64 bits for  _any_ currently representable ledger, where the transaction index is less than 65536.

In crude ASCII, the proposed format looks like this:

```
     +-------------------------------------+
     |                                     |
     | +-----------+                       |
     | |           |                       |
     | |           v                       v
1RRRRTTL CCCCCCCC [AAAAAAAA ... AAAAAAAA] [BBBBBBBB ... BBBBBBBB ]
     | | |
     | | |
     | | +---> Checksum, as described by you originally.
     | |
     | +---> Ledger Index Length: `1` if ledger index is 64-bits long, `0` otherwise.
     |
     +---> Transaction Index Length: number of bytes _minus one_ occupied by transaction index.
```

This will fit in 64 `unsigned` bits for _any_ ledger where the transaction index is less than 65536. Considering that the existing format can't represent such transactions anyways, and we're comfortable with that, we should be fine. Bonus: 4 reserved bits (potentially 5 if we don't care for a leading 1). How to use them? There's options. For example, we could shave off two of the "reserved" bits to represent `mainnet` (`0`), `testnet` (`1`) and `devnet` (`2`) and use `3` to represent _other_.

Small caveat: we need a simple rule to ensure there is a canonical format (that is, the same tx can't be represented by two CTIs):

1. If the ledger index is 64-bits long but the encoded value is less than 2<sup>32</sup>-1 the CTI is invalid.
2. If the transaction index has _leading_ zero bytes (i.e. is not represented minimally) the CTI is invalid.

### Example:

Your example was CTI  resides in ledger 62084722, has a transaction index of 25, the transaction hash's first nibble is 1 and the ledger hash's first nibble is F.

Encoded with my scheme, would produce:

`0x80` for the flag bits, `0` for the transaction index indicator (it's 25 which needs 1 byte) and `0` for the ledger length (it's 32 bits): the result is `0x80`.
The checksum is, per your calculations, `0x1F`.
The ledger index is, per your calculations, `0x03B35672`.
The transaction index is, per your calculations, `0x19`

Let's compare the CTIs produced by the two encodings:
| Style 	|       Hex      	|    CTI    	|
|-------	|:----------------:	|:-----------------:	|
| NB  	| `0x801F03B3567219` 	| **`36062897773113881`** |
| RH   	| `0xF1001903B35672` 	| **`67835576823535218`** |

#### Downside
Once parsed into a 64-bit `unsigned` value, the location of the "header" is 'floating' (i.e. it's not always in the high byte, hence the need for the leading one in the header). This makes parsing a _little_ more difficult; the cost for having ever so slightly more compact CTIs. This may or may not be a worthwhile tradeoff.

This downside can be addressed by ensuring that the transaction index is always represented as two bytes. Of course, this makes the "overlong" checker slightly more complex. Still, it may be worth the effort.

#### Representation
Decimal. I still think hex is, generally, more compact and (imo) easier to parse, but I think decimal is the way to go for the reasons @RichardAH highlights.

**Reply by nbougalis** [reply]
There's a couple of issues with endianess in my example that I'll need to work out.

**Comment by nbougalis** [comment]
@RichardAH liked the proposed extensions to the format, but he also noticed that with a mild massaging of the fields, my proposal could be backwards-compatible with existing CTIs. He's going to make those changes in the proposed standard.

A win, all-around!

**Reply by RichardAH** [reply]
XLS15.1 update has been mostly made

**Reply by interc0der** [reply]
Here is my shot at the advanced implementation... How does it look @RichardAH ?

### Implementation

```ts
  encode: (opts: CtiAdvancedEncodeOpts) => {
    let ledger_check = BigInt(parseInt(opts.ledger_hash.slice(0, 1), 16));
    let txn_check = BigInt(parseInt(opts.txn_hash.slice(0, 1), 16));
    let txnBin = Number(opts.txn_index).toString(2);
    let lgrBin = Number(opts.ledger_index).toString(2);

    let padding = BigInt('00'); // Will be dropped from BigInt, placeholder
    let T = BigInt(txnBin.length < 16 ? 0 : 1);
    let L = BigInt(lgrBin.length < 32 ? 0 : 1);

    let cti = (padding << 1n) + T;
    cti <<= 1n;
    cti += L;
    cti <<= 4n;
    cti += BigInt(opts.networkId);
    cti <<= 4n;
    cti += ledger_check;
    cti <<= 4n;
    cti += txn_check;
    cti <<= T === 1n ? 32n : 16n;
    cti += BigInt(opts.txn_index);
    cti <<= L === 1n ? 64n : 32n;
    cti += BigInt(opts.ledger_index);
    return cti;
  },
```

### Example

| | |
| --- | ----------- |
| networkId      | 1       |
| ledger_hash   | F8A87917637D476E871D22A1376D7C129DAC9E25D45AD4B67D1E75EA4418654C       |
|   ledger_index  | 62084722       |
| txn_hash   | 25       |

| | |
| --- | ----------- |
|hex| 0x1F1001903B35672|
| ctx   | 139893170861463154|

### Question

Since there are floating zeros at the front end of the bits, these will get chopped off when encoding. This is especially an issue when trying to determine the T & L bits when they are both 0 ( TL = 00 ). 

Is there a simple way to decode the advanced CTI? I would imagine that the first step would be to determine the T & L bits and then move from there. However, in their current position, they could be lost and inaccessible.

**Comment by mvadari** [comment]
@RichardAH should this spec be moved into PR and kept in the repo for historical/record-keeping purposes?
